lib/read-theme.js

```diff
@@ -105,16 +108,84 @@ processHelpers = function (theme, themeFile) {
     }
 };
 
-module.exports = function readTheme(themePath) {
-    return readThemeStructure(themePath).then(function (themeFiles) {
-        var theme = {
-            path: themePath,
-            files: themeFiles,
-            results: {
-                pass: [],
-                fail: {}
+/**
+ * Works only for posts, pages and custom templates at the moment.
+ */
+extractCustomTemplates = function extractCustomTemplates(allTemplates) {
+    var toReturn = [],
+        generateName = function generateName(templateName) {
+            var name = templateName;
+
+            name = name.replace(/^post-|page-|custom-/, '');
+            name = name.replace(/-/g, ' ');
+            name = name.replace(/\b\w/g, function (letter) {
+                return letter.toUpperCase()
+            });
+
+            return name.trim();
+        },
+        generateFor = function (templateName) {
+            if (templateName.match(/^page-/)) {
+                return ['page'];
+            }
+
+            if (templateName.match(/^post-/)) {
+                return ['post'];
             }
+
+            return ['page', 'post'];
+        },
+        generateSlug = function (templateName) {
+            if (templateName.match(/^custom-/)) {
+                return null;
+            }
+
+            return templateName.match(/^(page-|post-)(.*)/)[2];
         };
-        return readHbsFiles(theme);
+
+    _.each(allTemplates, function (templateName) {
+        if (templateName.match(/^post-|page-|custom-/)) {
+            toReturn.push({
+                filename: templateName,
+                name: generateName(templateName),
+                'for': generateFor(templateName),
+                slug: generateSlug(templateName)
+            });
+        }
     });
+
+    return toReturn;
+};
+
+/**
+ * Extracts from all theme files the .hbs files.
+ */
+extractTemplates = function extractTemplates(allFiles) {
+    return _.reduce(allFiles, function (templates, entry) {
+        var tplMatch = entry.file.match(/(^[^\/]+).hbs$/);
+        if (tplMatch) {
+            templates.push(tplMatch[1]);
+        }
+        return templates;
+    }, []);
+};
+
+module.exports = function readTheme(themePath) {
+    return readThemeStructure(themePath)
+        .then(function (themeFiles) {
+            var allTemplates = extractTemplates(themeFiles);
+
+            return readHbsFiles({
+                path: themePath,
+                files: themeFiles,
+                templates: {
+                    all: allTemplates,
```

Comment:
See https://github.com/TryGhost/Ghost/blob/master/core/server/themes/active.js#L45
It's useful for Ghost to return both: all templates and the custom templates.